### PR TITLE
Bounty#140

### LINKED ISSUE #140: Remove wallet can only submit once logic from NFT Gated Deals
if (!nftCollectionRules.purchaseAmountPerToken && nftCollectionRules.purchaseAmount > 0) {
                require(!_data.nftWalletUsedForPurchase[_collectionAddress][msg.sender], "wallet already used for nft set");

this logic in AelinNFTGating.sol is something that we have never used and the logic does not make sense. We should remove it from the logic set. There is the option to have it so that a wallet can only participate once and all their NFTs become blacklisted. the reason it doesn't make sense is we need to pass in the token ids to be blacklisted. if a pool is used with this setting it just incentivizes people to send their NFTs to different wallets which is not ideal. we should just remove it

### PR DESCRIPTION
**Pull request by CookedCookee**
Closes #140 

1. Removed the if block from the purchaseDealTokensWithNft function, and the wallet struct variable from the gating library contract
2. Removed the analogous logic from the legacy pool contracts
3. Removed needless complexity from the NftCollectionRules struct with respect to the per wallet logic also

I hope these changes were what you were looking for. If not, let me know and I'll make further adjustments as needed.

All the best,
Cooki

### CODE REVIEW

File: contracts/AelinPool.sol [hunk 1]
```diff
@@ -236,11 +231,7 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
             NftCollectionRules memory nftCollectionRules = nftCollectionDetails[collectionAddress];
             require(nftCollectionRules.collectionAddress == collectionAddress, "collection not in the pool");
 
-            if (nftCollectionRules.purchaseAmountPerToken) {
```
Comment (alextheboredape) [inline contracts/AelinPool.sol:239]: this is not right. we want to keep the following logic with a minor change:

if (nftCollectionRules.purchaseAmount > 0) {
    maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount * tokenIds.length;
}

and we want to remove this entire if statement. we are no longer using the nftWalletUsedForPurchase mapping anymore:

            if (!nftCollectionRules.purchaseAmountPerToken && nftCollectionRules.purchaseAmount > 0) {
                require(!nftWalletUsedForPurchase[collectionAddress][msg.sender], "wallet already used for nft set");
                nftWalletUsedForPurchase[collectionAddress][msg.sender] = true;
                maxPurchaseTokenAmount += nftCollectionRules.purchaseAmount;
            }

Comment (CookedCookee) [reply contracts/AelinPool.sol:239]: Oh yes, sorry my mistake

File: package.json [hunk 2]
```diff
@@ -76,7 +76,7 @@
 		"@nomiclabs/hardhat-ethers": "2.0.2",
```
Comment (0xlinus) [inline package.json:76]: Please remove this. Only add changes that are relevant to the PR

File: test/tests/AelinUpFrontDeal.purchase.t.sol [hunk 3]
```diff
@@ -1397,11 +1316,7 @@ contract AelinUpFrontDealPurchaseTest is Test, AelinTestUtils, IAelinUpFrontDeal
         emit AcceptDeal(user2, purchaseCollection2, purchaseCollection2, poolSharesAmount, poolSharesAmount);
         AelinUpFrontDeal(dealAddressNftGating721).acceptDeal(nftPurchaseList, merkleDataEmpty, purchaseCollection2);
 
-        // user2 is now blacklisted from purchasing again in this collection
-        vm.expectRevert("wallet already used for nft set");
-        AelinUpFrontDeal(dealAddressNftGating721).acceptDeal(nftPurchaseList, merkleDataEmpty, purchaseCollection2);
-
-        // case 3: [collection1] user2 comes back and max out their allocation (purchaseAmountPerToken = true)
+        // case 2: [collection1] user2 comes back and max out their allocation (purchaseAmountPerToken = true)
```
Comment (0xlinus) [inline test/tests/AelinUpFrontDeal.purchase.t.sol:1319]: nitpicking: case 3?

File: package.json (hunk 2 above)
Comment (alextheboredape) [reply package.json:76]: I think its ok. we are only updating openzeppelin contracts. the rest is moving packages around to re-alphabatize them.

File: test/tests/AelinPool.misc.t.sol [hunk 4]
```diff
@@ -1078,7 +1077,7 @@ contract AelinPoolMiscTest is Test, AelinTestUtils {
         );
         assertEq(
             AelinDeal(poolVars.dealAddress).maxTotalSupply(),
-            boundedVars.purchaseTokenTotalForDeal * 10**(18 - purchaseToken.decimals()),
+            boundedVars.purchaseTokenTotalForDeal * 10 ** (18 - purchaseToken.decimals()),
```
[the remark this answers is not in the document]
Comment (alextheboredape) [reply test/tests/AelinPool.misc.t.sol:1080]: cb has the same compiler change. We should add this to the global linting settings so my computer does the same style. I think its ok to leave it

**Review by alextheboredape — APPROVED** [review]

File: package-lock.json
Comment (0xcdb) [inline package-lock.json:1]: @CookedCookee can you please remove this change? We can't have anyone external update package-lock.json

File: package.json (hunk 2 above)
Comment (0xcdb) [reply package.json:76]: @CookedCookee please remove the change for oppenzeppelin except if it's needed for a specific reason?

File: test/tests/AelinUpFrontDeal.purchase.t.sol [hunk 5]
```diff
@@ -1397,10 +1316,6 @@ contract AelinUpFrontDealPurchaseTest is Test, AelinTestUtils, IAelinUpFrontDeal
         emit AcceptDeal(user2, purchaseCollection2, purchaseCollection2, poolSharesAmount, poolSharesAmount);
         AelinUpFrontDeal(dealAddressNftGating721).acceptDeal(nftPurchaseList, merkleDataEmpty, purchaseCollection2);
 
-        // user2 is now blacklisted from purchasing again in this collection
-        vm.expectRevert("wallet already used for nft set");
```
Comment (0xcdb) [inline test/tests/AelinUpFrontDeal.purchase.t.sol:1401]: @CookedCookee when it's an  ERC721 collection I guess we can replace this verification with `tokenId already used` instead as we want to make sure users can't buy again with the tokenIDs they just used

File: test/tests/AelinUpFrontDeal.purchase.t.sol [hunk 6]
```diff
@@ -1560,18 +1414,14 @@ contract AelinUpFrontDealPurchaseTest is Test, AelinTestUtils, IAelinUpFrontDeal
...
 
         // user1 now purchases half of the allocation
         uint256 purchaseAmount = purchaseCollection / 2;
         uint256 poolSharesAmount = (purchaseAmount * 10 ** underlyingTokenDecimals) / purchaseTokenPerDealToken;
         vm.expectEmit(true, false, false, true);
         emit AcceptDeal(user1, purchaseAmount, purchaseAmount, poolSharesAmount, poolSharesAmount);
         AelinUpFrontDeal(dealAddressNftGatingPunks).acceptDeal(nftPurchaseList, merkleDataEmpty, purchaseAmount);
-
-        // user1 now tries to buy the remaining allocation, but fails
-        vm.expectRevert("wallet already used for nft set");
```
Comment (0xcdb) [inline test/tests/AelinUpFrontDeal.purchase.t.sol:1573]: same here

File: test/tests/AelinUpFrontDeal.purchase.t.sol [hunk 7]
```diff
@@ -1585,202 +1435,19 @@ contract AelinUpFrontDealPurchaseTest is Test, AelinTestUtils, IAelinUpFrontDeal
         vm.expectEmit(true, false, false, true);
         emit AcceptDeal(user2, purchaseCollection, purchaseCollection, poolSharesAmount, poolSharesAmount);
         AelinUpFrontDeal(dealAddressNftGatingPunks).acceptDeal(nftPurchaseList, merkleDataEmpty, purchaseCollection);
-
-        // user2 now tries to buy some extra tokens, but fails
-        vm.expectRevert("wallet already used for nft set");
```
Comment (0xcdb) [inline test/tests/AelinUpFrontDeal.purchase.t.sol:1590]: Same

**Review by 0xcdb — CHANGES_REQUESTED** [review]
@CookedCookee great stuff! A few minor comments to address before I can merge :)

**Review by 0xcdb — APPROVED** [review]

**Pull request merged**